exchanges/orderbook/linked_list.go

```diff
@@ -0,0 +1,619 @@
+	// back on to the stack.
+	old := *tip
+	switch {
+	case old.prev == nil: // At head position
+		// shift current tip head to the right
+		*tip = old.next
+		// Remove reference to node from chain
+		if old.next != nil { // This is when liquidity hits zero
+			old.next.prev = nil
+		}
+	case old.next == nil: // At tail position
+		// Remove reference to node from chain
+		old.prev.next = nil
+	default:
+		// Reference prior node in chain to next node in chain
+		// bypassing current node
+		old.prev.next = old.next
+		old.next.prev = old.prev
+	}
+	ll.length--
+	return old
+}
+
+// insertAtTip inserts at a tip target (can inline)
+func insertAtTip(ll *linkedList, tip **node, updt Item, stack *stack) {
+	n := stack.Pop()
+	n.value = updt
+	n.next = *tip
+	n.prev = (*tip).prev
+	if (*tip).prev == nil { // Tip is at head
+		// Replace head which will push everything to the right
+		// when this node will reference new node below
+		*tip = n
+	} else {
+		// Reference new node to previous node
+		(*tip).prev.next = n
+	}
+	// Reference next node to new node
+	n.next.prev = n
+	ll.length++
+}
+
+// insertAtTail inserts at tail end of node chain (can inline)
+func insertAtTail(ll *linkedList, tip **node, updt Item, stack *stack) {
+	n := stack.Pop()
+	n.value = updt
+	// Reference tip to new node
+	(*tip).next = n
+	// Reference new node with current tip
+	n.prev = *tip
+	ll.length++
+}
+
+// insertHeadSpecific inserts at head specifically there might be an instance
+// where the liquidity on an exchange does fall to zero through a streaming
+// endpoint then it comes back online. (can inline)
+func insertHeadSpecific(ll *linkedList, updt Item, stack *stack) {
+	n := stack.Pop()
+	n.value = updt
+	ll.head = n
+	ll.length++
+}
+
+// insertNodeAtBookmark inserts a new node at a bookmarked node position
+// returns if a node needs to replace head (can inline)
+func insertNodeAtBookmark(ll *linkedList, bookmark, n *node) {
+	switch {
+	case bookmark == nil: // Zero liquidity and we are rebuilding from scratch
+		ll.head = n
+	case bookmark.prev == nil:
+		n.prev = bookmark.prev
+		bookmark.prev = n
+		n.next = bookmark
+		ll.head = n
+	case bookmark.next == nil:
+		n.prev = bookmark
+		bookmark.next = n
+	default:
+		bookmark.prev.next = n
+		n.prev = bookmark.prev
+		bookmark.prev = n
+		n.next = bookmark
+	}
+	ll.length++
+}
+
+// shiftBookmark moves a bookmarked node to the tip position or if nil sets
+// tip as bookmark (can inline)
+func shiftBookmark(tip *node, bookmark, head **node, updt Item) bool {
+	if *bookmark == nil { // End of the chain and no bookmark set
+		*bookmark = tip // Set tip to bookmark so we can set a new node there
+		return false
+	}
+	(*bookmark).value = updt
+	(*bookmark).next.prev = (*bookmark).prev
+	if (*bookmark).prev == nil { // Bookmark is at head
+		*head = (*bookmark).next
+	} else {
+		(*bookmark).prev.next = (*bookmark).next
+	}
+	tip.next = *bookmark
+	(*bookmark).prev = tip
```

Comment:
These lines overwrite what you've just done in 613.
bookmark.prev.next = bookmark.next
tip.next = bookmark
bookmark.prev = tip
.*. bookmark.prev.next  = bookmark
then finally you set bookmark.next = nil

This makes it seem like a different scenario. I may need a summary of what's going on here

Reply:
`ln: 613` removes the reference of the current bookmarked node in the chain.
`tip.next = *bookmark` inserts the bookmark after the tip node
`bookmark.prev = tip` reassigns the previous relationship with the current tip 
 `bookmark.next = nil` because we are at the tail end of the chain we nill this reference out
`.*. bookmark.prev.next = bookmark`  != `bookmark.prev.next = bookmark.next`

Have you come across some problems in testing?

Reply:
Yeah, this function doesn't have a test, so I had to write one:
```

func TestShiftBookmark(t *testing.T) {
	prev := &node{
		value: Item{
			ID: 1336,
		},
		next:    nil,
		prev:    nil,
		shelved: time.Time{},
	}
	next := &node{
		value: Item{
			ID: 1338,
		},
		next:    nil,
		prev:    prev,
		shelved: time.Time{},
	}
	bookmark := &node{
		value: Item{
			ID: 1337,
		},
		next:    next,
		prev:    prev,
		shelved: time.Time{},
	}
	tip := &node{
		value: Item{
			ID: 69420,
		},
		next:    next,
		prev:    prev,
		shelved: time.Time{},
	}
	updt := Item{
		ID: 1339,
	}
	prev.next = bookmark
	next.prev = bookmark
	shiftBookmark(tip, &bookmark, nil, updt)
}

```

The line `	(*bookmark).prev = tip` does still overwrite what happens in 
```
 else {
		(*bookmark).prev.next = (*bookmark).next
	}
```
But maybe you can help with clarifying how I've structured my test (and then please add one :D)

Reply:
Added tests for this.